README.md
# natlas

natlas - Network Atlas  
Michael Laforest `<mjlaforest` *at* `gmail` *dot* `com>`

Automated discovery and diagram tools using SNMP, CDP, and LLDP.

 ```# ./natlas-cli.py diagram -r 10.75.0.1 -o .\network.svg```<br><br><br>*The above command will generate the diagram to the right.* | ![natlas-Diagram Main][diag-main]
:--- | --- 
 
# Support

If you use any of these tools or find them useful please consider donating.  

Donation Method | Address | QR Code
--- | --- | ---
Bitcoin Cash (BCH) | 1HSycjR3LAZxuLG34aEBbQdUSayPkh8XsH | ![1HSycjR3LAZxuLG34aEBbQdUSayPkh8XsH](https://raw.github.com/MJL85/natlas/master/docs/donate/BCH.png "Bitcoin Cash (BCH)")
Bitcoin (BTC) | 1HY3jPYVfE6YZbuYTYfMpazvSKRXjZDMbS  | ![1HY3jPYVfE6YZbuYTYfMpazvSKRXjZDMbS](https://raw.github.com/MJL85/natlas/master/docs/donate/BTC.png "Bitcoin (BTC)")

# About

natlas is a python application framework which includes a discovery engine, an API, and a front-end cli for running available modules.

##### For Users
The `natlas-cli.py` program allows you to browse and execute any natlas module, including, among others, the network discovery and diagram module.

##### For Developers

Copy `docs/template_module.py` to a new file in the `modules/` directory.
Edit the function `mod_load()` to set the name and help properties for your module.
The entry function for any module is `mod_entry()`.
When creating a new module, natlas will create a new object and pass it to mod_entry().  From there the natlas API is available and includes functions such as:
- discover_network()
- query_node()
- get_switch_vlans()

Once saved, your module will automatically be listed in `natlas-cli.py list` and runnable.

# Modules
| Module | Description |
| --- | --- |
| diagram | Discovers a network and generates a diagram based on CDP and LLDP neighbor information. |
| get-mac-table | Collect a list of all MAC addresses on the discovered network and generate a report. |
| get-arp-table | Collect a list of all ARP entries. |
| get-hosts | Determine all hosts connected to one or all switches in the network. Includes MAC, IP, DNS name of each host, along with what switch and port it was found on. Can be exported to CSV. |
| tracemac | Trace a MAC address through a layer 2 network. |
| newconfig, showconfig, checkconfig | Modules to create, display, and validate natlas configuration files. |

# Network Discovery  
  
The discovery process uses SNMP, CDP, and LLDP to discover the network topology and details about each node.  Each discovered node will be evaluated against the `discover` ACL (defined in the config file) to determine how to proceed; the ACL may allow discovery, stop discovery here, or include it as a leaf in the diagram.

<table>
	<tr>
	<td valign=top rowspan=2>
		The <i>discover</i> ACL is defined as
	<pre><code>"discover" : [
	ACE1,
	ACE2,
	...
	ACEn,
]</code></pre>
	</td>
	<td valign=top>
		An <i>ACE</i> is defined as  
	<pre><code>&lt;permit|deny|leaf|include|;&gt; &lt; [host REGEX] | [ip CIDR] &gt;</code></pre>
	</td>
	</tr>
	<tr>
	<td>
Example
	<pre><code>"discover" : [
	"deny ip 10.50.12.55",
	"deny host ^SEP.*",
	"permit ip 10.50.12.0/24",
	"leaf host ^Switch2$",
	"permit ip any"
]</code>	</td>
	</tr>
</table>

---

| ACE Match Type| Include Node | Collect Node Information | Allow Discovery of Adjacencies |
| --- |:---:|:---:|:---:|
| **permit** | yes | yes | yes |
| **leaf** | yes | yes |  |
| **include** | yes |  |  |
| **deny** |  |  |  |

  ---
 
| ACE Parameter | Description | Example |
| --- | --- | --- |
| host REGEX | The host can match any regular expression string.  The host string is what is reported from CDP or LLDP. | `host Router-.*` |
| ip CIDR | The ip can be matched against and CIDR. | `ip 10\\.50\\.31\\.0/24` |
| platform REGEX | The system platform or hardware model. | `platform .*3850.*` |
| software REGEX | The system software version or IOS. | `software ^15\\.3` |

# Command Reference

### Diagram
```
# natlas-cli.py diagram -r <root IP>
                -o <output file>
               [-d <max depth>]
               [-c <config file>]
               [-t <diagram title>]
               [-C <catalog file>]
```
| Option | Description |
| --- | --- |
| `-r <root IP>` | IP address of the network node to start on. |
| `-o <output file>` | The file that the output will be written to.<br />Common file extensions: `.png`, `.pdf`, `.svg` |
| `-c <config file>` | The JSON configuration file to use. |
| `-d <max depth>` | The maximum hop depth to discover, starting at the root node specified by `-r` |
| `-t <diagram title>` | The title to give your generated network diagram. |
| `-C <catalog file>` | If specified, natlas will generate a comma separated (CSV) catalog file with a list of all devices discovered. |

### get-mac-table
```
# natlas-cli.py get-mac-table -n <node IP> -c <snmp v2 community> [-m <mac regex>] [-p <port regex>] [-v <vlan>]
```
| Option | Description |
| --- | --- |
| `-n <node IP>` | IP address of the network node to get MAC addresses from. |
| `-c <community>` | SNMPv2 community string. |
| `-m <regex>` | Filter MAC addresses by this regex string. |
| `-p <regex>` | Filter MAC addresses to only those on ports matching this regex string. |
| `-v <regex>` | Filter MAC addresses to only those on VLANs matching this regex string. |

### get-arp-table
```
# natlas-cli.py get-arp-table -n <node IP> -c <snmp v2 community> [-s | -d] [-i <IP regex>] [-v <vlan regex>] [-m <mac regex>]
```
| Option | Description |
| --- | --- |
| `-n <node IP>` | IP address of the network node to get ARP entries from. |
| `-c <community>` | SNMPv2 community string. |
| `-s` | Include static entries only |
| `-d` | Include dyntamic entries only |
| `-i <regex>` | Include entries with IP addresses that match regex pattern |
| `-v <regex>` | Include entries with VLANs that match regex pattern |
| `-m <regex>` | Include entries with MAC addreses that match regex pattern |

### get-hosts

get-hosts can either collect information from a single node or can do a network discovery and collect information from all discovered nodes.
```
# natlas-cli.py get-hosts -r <root IP> -c <config file> [-o <csv file>] [-d <discovery depth>]

# natlas-cli.py get-hosts -n <node IP> [-r <router IP>] -C <snmp v2 community> [-v <vlan regex>] [-p <port regex>] [-o <csv file>]
```
| Option | Description |
| --- | --- |
| `-r <root IP>` | IP address to begin a network discovery. |
| `-c <config file>` | natlas configuration file to use. |
| `-o <csv file>` | Output CSV file path. |
| `-d <depth>` | Maximum network discovery depth. |
| --- | --- |
| `-n <node IP>` | IP address of single layer2 or layer3 node to collect information from. |
| `-r <router IP>` | IP address of the layer3 device to collect ARP entries from. If this is omitted then the IP from -n will be used. |
| `-C <community>` | SNMPv2 community string. |
| `-v <regex>` | Include entries with VLANs that match regex pattern |
| `-p <regex>` | Include entries on ports that match regex pattern |
| `-o <csv file>` | Output CSV file path. |

### tracemac

```
# natlas-cli.py tracemac -n <starting node IP> -m <MAC address>
```
| Option | Description |
| --- | --- |
| `-n <starting node IP>` | IP address of node to begin layer 2 MAC trace. |
| `-m <MAC address>` | MAC address to locate in the network. |

### Config
| | |
| --- | --- |
| `# natlas-cli.py newconfig` | Generate a new config file |
| `# natlas-cli.py showconfig [-c <config file>]` | Display the config file |
| `# natlas-cli.py checkconfig [-c <config file>]` | Validate the contents of the config file |

# Configuration File
The configuration file defines common parameters in a JSON format.
```
{
	"snmp" : [
		{ "community":"private",	"ver":2 },
		{ "community":"public",		"ver":2 }
	],
	"domains" : [
		".company.net",
		".company.com"
	],
	"discover" : [
		"permit ip 10.0.0.0/8",
		"permit host Router[1,2]",
		"deny ip any",
	],
 	"diagram" : {
		"node_text_size" : 10,
		"link_text_size" : 9,
		"title_text_size" : 15,
		"get_stack_members" : 0,
		"get_vss_members" : 0,
		"expand_stackwise" : 0,
		"expand_vss" : 0,
		"expand_lag" : 1,
		"group_vpc" : 0
    }
	
}
```

| Block / Variable | Description |
| --- | --- |
| `snmp` | Defines a list of SNMP credentials.  When connecting to a node, each of these credentials is tried in order until one is successful. |
| `discover` | Defines a Cisco-style ACL. See the `Network Discovery` section. |
| `diagram` | Defines values used by the diagram module.  Detailed below in the *Diagram block* table. |

### Diagram block
| Variable | Type | Default Value | Description |
| --- | --- | --- | --- |
| `node_text_size` | integer | `10` | Node text size. |
| `link_text_size` | integer | `9` | Link text size. |
| `title_text_size` | integer | `15` |  Diagram title text size. |
| `get_stack_members` | bool | `0` | If set to `1`, nodes will include details about stackwise members. |
| `get_vss_members` | bool | `0` | If set to `1`, nodes will include details about VSS members. |
| `expand_stackwise` | bool | `0` | If set to `1`, nodes belonging to stackwise groups will be expanded to show each member as a node. |
| `expand_vss` | bool | `0` | If set to `1`, nodes belonging to VSS groups will be expanded to show each member as a node. |
| `expand_lag` | bool | `1` | If set to `1`, each link between nodes will be shown.  If set to `0`, links of the same logical link channel will be grouped and only the channel link will be shown. |
| `group_vpc` | bool | `0` | If set to `1`, VPC peers will be grouped together on the diagram, otherwise they will not be clustered. |

# Diagram
natlas will attempt to collect the following information and include it in the generated diagram:
+ All devices (via CDP and LLDP)
+ Interface names
+ IP addresses
+ VLAN memberships
+ Etherchannel memberships (LACP only)
+ Identify trunk links
+ Identify switched links
+ Identify routed links
+ BGP Local AS
+ OSPF Router ID
+ HSRP Virtual IP
+ HSRP Priority
+ VSS Domain
+ Stackwise membership
+ VPC peerlink information

#### Diagram Formatting
+ Nodes
  + Circle nodes represent layer 2 switches.
  + Diamond nodes represent layer 3 switches or routers.
  + If a node has multiple borders then either VSS or StackWise is enabled.
    + VSS - Will always have a double border.
    + StackWise - The number of borders denotes the number of switches in the stack.
  + If the configuration specifies, VSS/VPC/Stackwise nodes will be grouped in larger squares.
+ Links
  + Links are shown with arrowed lines.  The end with no arrow is the *parent* and the end with the arrow is the *child*, such that the arrangement is *parent*->*child*.
  + If a link says *P:gi0/1* , *C:gi1/4* then the parent node's connection is on port gi0/1 and the child node's connection is on port gi1/4.
  + If the link is part of an Etherchannel the etherchannel's interface name will also be shown.  Since an etherchannel interface is locally significiant, a *P:* and *C:* will also be shown if available.

# Examples

#### Diagram Example 1
![natlas-Diagram Ex1][diag-ex1]

#### Diagram Example 2
![natlas-Diagram Ex2][diag-ex2]

#### Diagram Example 3
![natlas-Diagram Ex3][diag-ex3]

#### get-hosts Example 1
![natlas-get-hosts Ex1][get-hosts-ex1]

#### tracemac Example 1
![natlas-tracemac Ex1][tracemac-ex1]

# FAQ

#### Q1 - My diagram is too large.  I only want to diagram part of my network.
Try changing the config `discover` ACL to narrow down the scope of your discovery.  You can explicitly deny CIDR's or host name regex patterns if you do not want them included in your diagram.  
  
#### Q2 - Where is the config file?
Create a new one with
`# natlas-cli.py newconfig > natlas.conf`

#### Q3 - I need a diagram with less proprietary information. Can I get one without IPs or serial numbers?
You can change the text inside each node by editing the config option `diagram\node_text`. Below is an example that would produce a minimal information diagram:

```
"diagram" : {
	node_text = '<font point-size="10"><b>{node.name}</b></font><br />{node.ios}<br />{node.plat}'
}	
``` 
#### Q4 - How can I remove Cisco VoIP phones from my diagram?
```
"discover" : [
	"deny host ^SEP.*$"
]
``` 

[diag-main]: https://raw.github.com/MJL85/natlas/master/docs/images/diagram_example.png "Diagram Main"
[diag-ex1]: https://raw.github.com/MJL85/natlas/master/docs/images/diagram_example1.png "Diagram Example 1"
[diag-ex2]: https://raw.github.com/MJL85/natlas/master/docs/images/diagram_example2.png "Diagram Example 2"
[diag-ex3]: https://raw.github.com/MJL85/natlas/master/docs/images/diagram_example3.png "Diagram Example 3"
[get-hosts-ex1]: https://raw.github.com/MJL85/natlas/master/docs/images/get-hosts_example1.png "get-hosts Example 1"
[tracemac-ex1]: https://raw.github.com/MJL85/natlas/master/docs/images/tracemac_example1.png "tracemac Example 1"




CHANGELOG
---------------------
natlas
Michael Laforest
mjlaforest@gmail.com

CHANGE LOG
--------------------

v0.11 -
	- Rebranded to natlas
	- Reorganized into framework with dynamically loadable modules
	- Added Python version check
	- Better error info when loading improperly formatted config files
	- Added ACL match criteria for discovery: platform, serial, software
	- FIX: Platform / IOS / Serial sometimes not captured if SNMP was not available but CDP/LLDP had it
	- FIX: SNMP timeout could cause crash

	modules
	- Renamed: config -> newconfig
	- Added: showconfig
	- Added: checkconfig
	- Added: get-mac-table
	- Added: get-arp-table
	- Added: get-hosts
	- Added: tracemac
	- Removed: output_stdout

	API
	- Created the initial natlas API
	- Added: config_generate()
	- Added: config_validate()
	- Added: config_load()
	- Added: snmp_add_credential()
	- Added: set_discover_maxdepth()
	- Added: set_verbose()
	- Added: discover_network()
	- Added: new_node()
	- Added: query_node()
	- Added: write_diagram() 
	- Added: write_catalog() 
	- Added: get_switch_vlans()
	- Added: get_switch_macs()
	- Added: get_discovered_nodes()	
	- Added: get_node_ip()
	- Added: get_arp_table()
	- Added: get_neighbors()

v0.10 - 5/5/2018
	- Ported to Python 3
	- Cleaned up and refactored code
	- Improved discovery
	- Improved discovery console output
	- Fixed VSS chassis detection; now finds correct serial# and platform
	- All output referencing node IP now uses best IP instead of first found
	- Try all known IPs for a node until one works; no longer fails on unreacable IPs (eg, VRFs/ACL blocks/etc)
	- Default depth is now 100
	- Added runtime to stdout
	- Renamed 'graph' to 'diagram'

	- Cisco ACL-style node matching (replaced config subnets/exclude with discover)
	  - Added 'leaf' option to stop discovery beyond a matching node
	  - Added 'include' option to stop discovery at a matching node

	- Config option diagram/node_text replaces the below:
	  - include_svi
	  - include_lo
	  - include_serials

	diagram
	- Changed -f option to -o
	- If a LAG spans multiple devices (eg, Nexus) override expand_lag for that device
	- Output can create multiple files. Ex: -o "file.{svg|png}" will create file.svg and file.png

	getmacs
	- Added as new module

v0.9 -
	graph
	- Fixed bug with allowed VLAN lists not showing correct ranges.
	- Fixed bug where certain SNMP lookups could return empty sets and cause an exception 

v0.8 - 9/21/2015
	- Internal code changes.
	- Trunks show native VLAN (if mismatched then both P and C will be displayed)
	- Trunks show allowed VLANs (if mismatched then both P and C will be displayed) (standard range only)
	- Added support for LLDP
	- Renamed config option "collapse_stackwise" to "expand_stackwise"
	- Renamed config option "collapse_vss" to "expand_vss"

v0.7 - 7/25/2015
	graph
	- Added config option "include_svi" to include SVI info for nodes
	- Added config option "include_lo" to include loopback info for nodes
	- Added config option "include_serials" to include serial numbers in diagram for nodes
	- Added config option "get_stack_members" to include additional stack member info (serials, mac's etc).
		This info is also included in the catalog option, with each stack member being a line item.
	- Added config option "get_vss_members" to include additional VSSS member info.
	- Added config option "collapse_stackwise" and "collapse_vss". When =0 clusters will be expanded.
	- Removed SNMP platform detection and now use CDP. This means a low depth may not get the root platform.

v0.6 - 7/19/2015
	graph
	- Changed from PyDot to PyDot2
	- Fixed bug with CDP neighbors that do not advertise an IP (reported by Kent Coble)
	- Added config file option to change font size for node text
	- Added config file option to change font size for link text
	- Changed internal crawl method
	- Improved platform identification
	- Added IOS version identification

	config
	- Added command line module option to output a default config file

v0.4 - 7/10/2015
	- Improved speed.
	- Cleaned up code base.
	- Made PyNetAddr optional (change USE_NETADDR=1 to USE_NETADDR=0 to disable)

v0.3 - 7/8/2015
	- Added -C option to write a catalog of the discovery to a csv file.
	- Runs for ~1/4 the time.

v0.2 - 7/7/2015

	- Config: Changed "ignore" to "exclude". Now a list of CIDR's instead of specific IP addresses (requested by /u/LiftedKilt)
	- Config: Added "subnets" block.  List of CIDR's of allowed subnets to be crawled. (requeted by /u/1n5aN1aC)
	- Fixed bug related to unnumbered layer 3 interface (reported by /u/chazmosis)
	- Fixed bug related to Nexus host format (reported by /u/endemic)
	- Added OID for Nexus platform (thanks to /u/endemic)

v0.1 - 7/6/2015

	- Initial release


natlas/config.py
#!/usr/bin/python

'''
        natlas
        config.py

        Michael Laforest
        mjlaforest@gmail.com

        Copyright (C) 2015-2018 Michael Laforest

        This program is free software; you can redistribute it and/or
        modify it under the terms of the GNU General Public License
        as published by the Free Software Foundation; either version 2
        of the License, or (at your option) any later version.

        This program is distributed in the hope that it will be useful,
        but WITHOUT ANY WARRANTY; without even the implied warranty of
        MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
        GNU General Public License for more details.

        You should have received a copy of the GNU General Public License
        along with this program; if not, write to the Free Software
        Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
'''

import json
import sys

class natlas_config_diagram:
    node_text_size      = 8
    link_text_size      = 7
    title_text_size     = 15
    get_stack_members   = False
    get_vss_members     = False
    expand_stackwise    = False
    expand_vss          = False
    expand_lag          = True
    group_vpc           = False
    node_text           = '<font point-size="10"><b>{node.name}</b></font><br />' \
                            '{node.ip}<br />' \
                            '<%if {node.ios}: {node.ios}<br />%>' \
                            '<%if {node.plat}: {node.plat}<br />%>' \
                            '<%if ("{node.serial}"!=None)&({node.vss.enabled}==0)&({node.stack.enabled}==0): {node.serial}<br />%>' \
                            '<%if ({node.stack.enabled}==1)&({config.diagram.expand_stackwise}==1): {stack.serial}<br />%>' \
                            '<%if {node.vss.enabled}&({config.diagram.expand_vss}==1): {vss.serial}<br />%>' \
                            '<%if ({node.vss.enabled}==1)&({config.diagram.expand_vss}==0): VSS {node.vss.domain}<br />%>' \
                            '<%if {node.vss.enabled}&({config.diagram.expand_vss}==0): VSS 0 - {node.vss.members[0].plat} - {node.vss.members[0].serial}<br />VSS 1 - {node.vss.members[1].plat} - {node.vss.members[1].serial}<br />%>' \
                            '<%if {node.bgp_las}: BGP {node.bgp_las}<br />%>' \
                            '<%if {node.ospf_id}: OSPF {node.ospf_id}<br />%>' \
                            '<%if {node.hsrp_pri}: HSRP VIP {node.hsrp_vip}<br />HSRP Pri {node.hsrp_pri}<br />%>' \
                            '<%if {node.stack.enabled}: Stackwise {node.stack.count}<br />%>' \
                            '<%stack SW {stack.num} - {stack.plat} {stack.serial} ({stack.role})<br />%>' \
                            '<%loopback {lo.name} - {lo.ip}<br />%>' \
                            '<%svi VLAN {svi.vlan} - {svi.ip}<br />%>'

class natlas_discover_acl:
    '''
    Define an ACL entry for the 'discover' config block.
    Defined in the form:
        <action> <type> <str>
    Where
        <action> = permit, deny, leaf, nop
        <type>   = ip, host
        <str>    = string
    '''
    all_actions = [ ';', 'permit', 'deny', 'leaf', 'include' ]
    all_types   = [ ';', 'ip', 'host', 'software', 'platform', 'serial' ]

    def __init__(self, str):
        self.action     = "nop"
        self.type       = "nop"
        self.str        = "nop"

        t = list(filter(None, str.split()))
        if (len(t) < 3):
            raise Exception('Invalid ACL entry: "%s"' % str)

        self.action     = t[0]
        self.type       = t[1]
        self.str        = t[2]

        if (self.action not in self.all_actions):
            raise Exception('Invalid ACL entry: "%s"; %s' % (str, self.action))
        if (self.type not in self.all_types):
            raise Exception('Invalid ACL entry: "%s"; %s' % (str, self.type))

    def __repr__(self):
        return '<%s %s %s>' % (self.action, self.type, self.str)

class natlas_config:
    def __init__(self):
        self.host_domains       = []
        self.snmp_creds         = []
        self.discover_acl       = []
        self.diagram            = natlas_config_diagram()

    def load(self, filename):
        # load config
        json_data = self.__load_json_conf(filename)
        if (json_data == None):
            return 0

        self.host_domains       = json_data['domains']
        self.snmp_creds         = json_data['snmp']

        # parse 'discover' block ACL entries
        for acl in json_data['discover']:
            try:
                entry = natlas_discover_acl(acl)
            except Exception as e:
                print(e)
                return 0

            self.discover_acl.append(entry)

        json_diagram = json_data.get('diagram', None)
        if (json_diagram != None):
            self.diagram.node_text_size     = json_diagram.get('node_text_size', 8)
            self.diagram.link_text_size     = json_diagram.get('link_text_size', 7)
            self.diagram.title_text_size    = json_diagram.get('title_text_size', 15)
            self.diagram.get_stack_members  = json_diagram.get('get_stack_members', False)
            self.diagram.get_vss_members    = json_diagram.get('get_vss_members', False)
            self.diagram.expand_stackwise   = json_diagram.get('expand_stackwise', False)
            self.diagram.expand_vss         = json_diagram.get('expand_vss', False)
            self.diagram.expand_lag         = json_diagram.get('expand_lag', True)
            self.diagram.group_vpc          = json_diagram.get('group_vpc', False)
            self.diagram.node_text          = json_diagram.get('node_text', self.diagram.node_text)

        return 1

    def __load_json_conf(self, json_file):
        json_data = None
        fd = open(json_file)
        json_data = fd.read()
        fd.close()
        json_data = json.loads(json_data)
        return json_data

    def generate_new(self):
        return '{\n' \
                        '       "snmp" : [\n' \
                        '               { "community":"private", "ver":2 },\n' \
                        '               { "community":"public", "ver":2 }\n' \
                        '       ],\n' \
                        '       "domains" : [\n' \
                        '               ".company.net",\n' \
                        '               ".company.com"\n' \
                        '       ],\n' \
                        '       "discover" : [\n' \
                        '               "permit ip 10.0.0.0/8",\n' \
                        '               "permit ip 192.168.1.0/24",\n' \
                        '               "permit ip 0.0.0.0/32"\n' \
                        '       ],\n' \
                        '       "diagram" : {\n' \
                        '               "node_text_size" : 10,\n' \
                        '               "link_text_size" : 9,\n' \
                        '               "title_text_size" : 15,\n' \
                        '               "get_stack_members" : 0,\n' \
                        '               "get_vss_members" : 0,\n' \
                        '               "expand_stackwise" : 0,\n' \
                        '               "expand_vss" : 0,\n' \
                        '               "expand_lag" : 1,\n' \
                        '               "group_vpc" : 0\n' \
                        '       }\n' \
                        '}'


    def validate_config(self, filename):
        print('Validating config...')
        json_data = self.__load_json_conf(filename)
        if (json_data == None):
            print('Could not load config.')
            return 0

        ret = 0

        ret += self.__validate_config_snmp(json_data)
        ret += self.__validate_config_domains(json_data)
        ret += self.__validate_config_discover(json_data)
        ret += self.__validate_config_diagram(json_data)
            
        if (ret < 4):
            print('FAILED')
        else:
            print('PASSED')

    def __validate_config_snmp(self, data):
        sys.stdout.write('Checking snmp...')
        obj = None
        try:
            obj = data['snmp']
        except:
            print('does not exist')
            return 0

        if (type(obj) != list):
            print('not a list')
            return 0

        for cred in obj:
            if (type(cred) != dict):
                print('list contains a non-dict (%s)' % type(cred))
                return 0
            try:
                c = cred['community']
                if (type(c) != str):
                    print('community is not a string')
                    return 0
            except KeyError as e:
                print('one or more entries does not include %s' % e)
                return 0
            try:
                c = cred['ver']
                if (type(c) != int):
                    print('version is not an int')
                    return 0
                else:
                    if (c != 2):
                        print('version for \'%s\' is not supported' % cred['community'])
                        return 0
            except KeyError as e:
                print('one or more entries does not include %s' % e)
                return 0
        print('ok')
        return 1

    def __validate_config_domains(self, data):
        sys.stdout.write('Checking domains...')
        obj = None
        try:
            obj = data['domains']
        except:
            print('does not exist')
            return 0
        if (type(obj) != list):
            print('not a list')
            return 0
        for d in obj:
            if (type(d) != str):
                print('domain is not a string')
                return 0
        print('ok')
        return 1

    def __validate_config_discover(self, data):
        sys.stdout.write('Checking discover...')
        obj = None
        try:
            obj = data['discover']
        except:
            print('does not exist')
            return 0
        if (type(obj) != list):
            print('not a list')
            return 0
        for d in obj:
            if (type(d) != str):
                print('ACL is not a string')
                return 0

            ace = d.split(' ')
            if (len(ace) < 3):
                print('ACE not enough params \'%s\'' % d)
                return 0
            if (ace[0] not in natlas_discover_acl.all_actions):
                print('ACE op \'%s\' not valid' % ace[0])
                return 0
            if (ace[1] not in natlas_discover_acl.all_types):
                print('ACE cond \'%s\' not valid' % ace[1])
                return 0

        print('ok')
        return 1

    def __validate_config_diagram(self, data):
        sys.stdout.write('Checking diagram...')
        obj = None
        try:
            obj = data['diagram']
        except:
            print('does not exist')
            return 0
        if (type(obj) != dict):
            print('not a dict')
            return 0

        for nv in obj:
            if (nv not in ['node_text_size',
                            'link_text_size',
                            'title_text_size',
                            'get_stack_members',
                            'get_vss_members',
                            'expand_stackwise',
                            'expand_vss',
                            'expand_lag',
                            'group_vpc']):
                print('invalid value \'%s\'' % nv)
                return 0

        print('ok')
        return 1

